### DISCUSSION TITLE
Can it support the collection of cloud databases

### DISCUSSION THREAD

**Discussion by ghost** [burningalchemist/sql_exporter#517] [Q&A] [answered]
I understand whether the sql_exporter must be deployed together with the server where the database is located, But can I use this component for the cloud database I am using

**Comment by burningalchemist** [comment]
Hi @hunterzhang0909, yeah, it supports cloud databases too. Just make sure that your cloud database is accessible for sql_exporter.

I had no issues using sql_exporter with AWS or Google Cloud. I bet any other clouds should work just fine.

Do you have any specific database/cloud in mind? :)

**Reply by ghost** [reply]
hello, How can I use this component to complete the collection of 5000+ RDS instances? I cannot deploy it on the machine where the RDS is located，thanks

**Reply by ghost** [reply]
![66761716741013_ pic](https://github.com/burningalchemist/sql_exporter/assets/122633242/6e40cf95-8042-43a3-b54d-9ad93319dbac)

**Reply by burningalchemist** [reply]
Yeah, you can. 👍 Deploying is a not hard requirement, it's used in the config description just for the context.

You might want to use `jobs` configuration to specify your database connections. I haven't tested 5000 databases, though. 

Give it a shot and let me know. 🙂👍

**Reply by ghost** [reply]
okay, thank you very much

**Reply by ghost** [reply]
this is mysql.collector.yml：
collector_name: mysql
metrics:
  - metric_name: mysql_processlist
    type: gauge
    help: 'OS page file, used and available.'
    value_label: 'state'
    values: [id, user, host, db, command, state, time, sql_text, trx_time]
    query_ref: mysql_processlist_query
queries:
  - query_name: mysql_processlist_query
    query: |
      SELECT 
        pl.ID AS id,
        pl.USER AS user,
        SUBSTRING_INDEX(pl.HOST, ':', 1) AS host,
        pl.DB AS db,
        pl.COMMAND AS command,
        pl.TIME AS time,
        pl.STATE AS state,
        IFNULL(pl.INFO, '') AS sql_text,
        IFNULL(TIMESTAMPDIFF(SECOND, it.trx_started, CURRENT_TIMESTAMP), 0) AS trx_time
      FROM information_schema.processlist pl
      LEFT JOIN information_schema.innodb_trx it ON pl.id = it.trx_mysql_thread_id

**Reply by ghost** [reply]
Error: ts=2024-05-27T02:29:43.585Z caller=klog.go:134 level=error func=Errorf msg="Error gathering metrics: [from Gatherer #1] [collector=mysql,query=mysql_processlist_query] sql: Scan error on column index 1, name \"user\": converting driver.Value type []uint8 (\"admin\") to a float64: invalid syntax"

May I ask how to set field types

**Reply by burningalchemist** [reply]
Hi @hunterzhang0909,

Prometheus metric format is pretty strict ([link](https://prometheus.io/docs/concepts/data_model/#samples)): values have to be of `float64` type, the string values (e.g. `user`) can only exist as metric labels. 

So in your case, you need to make sure that all columns that do not return numbers are defined in the `key_labels` list. You also might want adjust your query to return less unnecessary columns.

I'd suggest starting with a small query, and then add more as needed. Also have a look at the examples [here](https://github.com/burningalchemist/sql_exporter/blob/master/documentation/sql_exporter.yml) and [here](https://github.com/burningalchemist/sql_exporter/blob/master/examples/mssql_standard.collector.yml). They are for MSSQL, but the syntax is the same.

If there is no particular metric value you'd like to extract (e.g. only collecting string data), you can use `static_value` parameter and set it to your desired value (say `1`).

**Reply by ghost** [reply]
![150191716805888_ pic](https://github.com/burningalchemist/sql_exporter/assets/122633242/a60940d1-316c-4e70-bc6e-669cdfbe7d0d)
I want to store connection information for database execution. Is it supported, json or string

**Reply by burningalchemist** [reply, marked as answer]
If you want to store any information in Prometheus, you have to follow its format (I posted the link above). So try maybe this snippet (change it accordingly to your requirements):

```
# Metric with a static value to retrieve string data.
- metric_name: mysql_data
  type: gauge
  help: 'Database server data'
  key_labels:
    # Populated from the columns of the result.
    - id
    - user
    - host
    - db
    - command
    - time
    - state
    - sql_text
    - trx_time
  # Static value, always set to `1`.
  static_value: 1
  query: |
    SELECT
    pl.ID AS id,
    pl.USER AS user,
    SUBSTRING_INDEX(pl.HOST, ':', 1) AS host,
    pl.DB AS db,
    pl.COMMAND AS command,
    pl.TIME AS time,
    pl.STATE AS state,
    IFNULL(pl.INFO, '') AS sql_text,
    IFNULL(TIMESTAMPDIFF(SECOND, it.trx_started, CURRENT_TIMESTAMP), 0) AS trx_time
    FROM information_schema.processlist pl
    LEFT JOIN information_schema.innodb_trx it ON pl.id = it.trx_mysql_thread_id
```

This is the best thing you can do if you don't actually need any data to be really counted/aggregated, and just want to store text - you have to store it in labels, there is no other way.
